### PR TITLE
Secret storage support

### PR DESCRIPTION
**Pull request by poljar**
Initial support for secret storage in the matrix-sdk-crypto crate.

This is the crypto crate part of: https://github.com/matrix-org/matrix-rust-sdk/issues/402

The main crate PR is at https://github.com/matrix-org/matrix-rust-sdk/pull/2621.

- [x] Public API changes documented in changelogs (optional)

### CODE REVIEW

File: crates/matrix-sdk-crypto/src/secret_storage.rs [hunk 1]
```diff
@@ -0,0 +1,787 @@
...
+    ///
+    /// This is uploaded to the homeserver in a global account data event.
+    #[zeroize(skip)]
+    storage_key_info: SecretStorageKeyEventContent,
+    /// The private key material.
+    secret_key: Box<[u8; 32]>,
+}
+
+impl std::fmt::Debug for SecretStorageKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```
Comment (Hywan) [inline crates/matrix-sdk-crypto/src/secret_storage.rs:135]: you can `use std::fmt` and then simplify this code to:

```suggestion
impl fmt::Debug for SecretStorageKey {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

Just a personal taste.

File: crates/matrix-sdk-crypto/src/secret_storage.rs [hunk 2]
```diff
@@ -0,0 +1,787 @@
...
+        f.debug_struct("SecretStorageKey")
+            .field("storage_key_info", &self.storage_key_info)
+            .finish()
```
Comment (Hywan) [inline crates/matrix-sdk-crypto/src/secret_storage.rs:138]:
```suggestion
            .finish_non_exhaustive()
```

Thoughts?

**Review by Hywan — COMMENTED** [review]
It looks excellent, well done!

File: crates/matrix-sdk-crypto/src/secret_storage.rs [hunk 3]
```diff
@@ -0,0 +1,787 @@
...
+        );
+
+        let key = Self { storage_key_info: key_info.to_owned(), secret_key: key };
+        key.check_zero_message()?;
+
+        Ok(key)
+    }
+
+    fn parse_base58_key(value: &str) -> Result<Box<[u8; 32]>, DecodeError> {
+        // Remove any whitespace we might have
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/secret_storage.rs:374]: The spec tells us so: https://spec.matrix.org/unstable/client-server-api/#recovery-key

> When reading in a recovery key, clients must disregard whitespace, and perform the reverse of steps 1 through 3.

[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/secret_storage.rs]: Added a comment.

File: crates/matrix-sdk-crypto/src/secret_storage.rs [hunk 4]
```diff
@@ -0,0 +1,787 @@
...
+    const ZERO_MESSAGE: &[u8; 32] = &[0u8; 32];
+    const PREFIX: [u8; 2] = [0x8b, 0x01];
+    const PREFIX_PARITY: u8 = Self::PREFIX[0] ^ Self::PREFIX[1];
+    const DEFAULT_KEY_ID_LEN: usize = 32;
+    #[cfg(not(test))]
+    const DEFAULT_PBKDF_ITERATIONS: u32 = 500_000;
+    #[cfg(test)]
+    const DEFAULT_PBKDF_ITERATIONS: u32 = 10;
+
+    // 35 bytes in total, 2 byte prefix, 32 bytes for the key material and one
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/secret_storage.rs]: My AI assistant says no to this one. I did change it to `32-byte prefix` though.

File: crates/matrix-sdk-crypto/src/secret_storage.rs [hunk 5]
```diff
@@ -0,0 +1,787 @@
...
+        rng.fill_bytes(key.as_mut_slice());
+
+        let key_id = Alphanumeric.sample_string(&mut rng, Self::DEFAULT_KEY_ID_LEN);
+
+        Self::from_bytes(key_id, key)
+    }
+
+    /// Create a new passphrase based [`SecretStorageKey`].
+    ///
+    /// The passphrase will be expanded into a 32 byte key using the `m.pbkdf2`
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/secret_storage.rs]: Same as the previous one.

File: crates/matrix-sdk-crypto/src/secret_storage.rs [hunk 6]
```diff
@@ -0,0 +1,787 @@
...
+                        .ok_or_else(|| DecodeError::MacLength(MAC_SIZE, mac.as_bytes().len()))?;
+
+                    key.verify_mac(&ciphertext, expected_mac.as_bytes())?;
+
+                    Ok(())
+                }
+            }
+            _ => {
+                // TODO: Expose the algorithm name from Ruma so we can put it
+                // into the Error.
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/secret_storage.rs:259]: https://github.com/ruma/ruma/issues/1661

**Review by Hywan — APPROVED** [review]

File: crates/matrix-sdk-crypto/src/secret_storage.rs [hunk 7]
```diff
@@ -0,0 +1,814 @@
...
+        Ok(plaintext)
+    }
+
+    /// The info about the [`SecretStorageKey`] formatted as a
+    /// [`SecretStorageKeyEventContent`].
+    ///
+    /// The [`SecretStorageKeyEventContent`] will contain information about this
+    /// secret storage key which then can be used to check if the user has
+    /// entered the correct key when trying to restore a [`SecretStorageKey`]
+    /// using the [`SecretStorageKey::from_account_data()`] method.
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/secret_storage.rs:527]:
```suggestion
    /// The [`SecretStorageKeyEventContent`] contains information about the
    /// secret storage key. This information can be used to determine whether
    /// the secret the user has entered is a valid secret for unlocking the
    /// Secrets Storage (i.e. a valid [`SecretStorageKey`]).
```

File: crates/matrix-sdk-crypto/src/secret_storage.rs [hunk 8]
```diff
@@ -0,0 +1,814 @@
...
+        &self.storage_key_info.key_id
+    }
+
+    /// The event type of this [`SecretStorageKey`].
+    ///
+    /// Can be used when uploading the key info as a
+    /// [`SecretStorageKeyEventContent`] to the homeserver.
+    ///
+    /// Will be a concatenation of `m.secret_storage.key.` and the key ID from
+    /// the [`SecretStorageKey::key_id()`] method.
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/secret_storage.rs:543]:
```suggestion
    /// The type is equal to the concatenation of the string
    /// `"m.secret_storage.key."` and the key ID from the
    /// [`SecretStorageKey::key_id()`] method.
```

**Pull request merged**